Solve bypass mem stage bug

vodogrey2012 opened the pull request:
Good evening. I have tried to solve bug https://github.com/MIPT-ILab/mipt-mips/issues/910.
I implemented data bypass logic similarly to the MEMORY stage, as you advised.

vodogrey2012 committed:
Split MEM and BRANCH bypass ports

Now there is the same Port error, but with BRANCH_2_EXECUTE_BYPASS port.

vodogrey2012 committed:
Add bypass to branch stage

vodogrey2012 committed:
Add comment about BRANCH bypass

pavelkryukov commented on simulator/modules/decode/bypass/data_bypass_interface.h:
```diff
@@ -19,19 +19,21 @@ class RegisterStage
     void set_to_mem_stage() { set_to_stage( MEM_STAGE); }
     void set_to_writeback() { set_to_stage( WB_STAGE); }
     void set_to_in_RF()     { set_to_stage( IN_RF_STAGE); }
+    void set_to_branch_stage() { set_to_stage( BRANCH_STAGE); }
```
This method is not used anywhere....

pavelkryukov commented on simulator/modules/decode/bypass/data_bypass_interface.h:
```diff
@@ -19,19 +19,21 @@ class RegisterStage
     bool is_mem_stage() const { return is_same_stage( MEM_STAGE); }
     bool is_writeback() const { return is_same_stage( WB_STAGE); }
     bool is_in_RF() const     { return is_same_stage( IN_RF_STAGE); }
+    bool is_branch_stage() const { return is_same_stage( BRANCH_STAGE); }
```
... so it always returns false...

pavelkryukov commented on simulator/modules/decode/bypass/data_bypass_interface.h:
```diff
@@ -72,6 +76,9 @@ class BypassCommand
         if ( bypassing_stage.is_mem_stage())
             bypass_direction = 2;
 
+        if ( bypassing_stage.is_branch_stage())
```
...and that `if` is not taken for all cases.

vodogrey2012 committed:
Add is_branch_stage_required() method

pavelkryukov commented on simulator/modules/core/perf_instr.h:
```diff
@@ -51,6 +51,9 @@ class PerfInstr : public FuncInstr
                                                 this->is_store() ||
                                                 this->is_jump()  ||
```
Probably, you have to exclude `this->is_jump()` from here.

vodogrey2012 replied:
It seems to me that it is better to remove ``this->is_indirect_jump()`` from ``is_branch_stage_required()``?

pavelkryukov replied:
I don't think so. First of all, it just contradicts with common sense — branch stage is required for branches, and not required for anything else. Additionally, MIPS has "branch-and-link" instructions, which store return address likewise `jal`

pavelkryukov commented on simulator/modules/decode/bypass/data_bypass_interface.h:
```diff
@@ -72,6 +76,9 @@ class BypassCommand
             bypass_direction = 2;
 
+        if ( bypassing_stage.is_branch_stage())
+                    bypass_direction = 4;
```
Please fix indentation here.

pavelkryukov commented on simulator/modules/decode/bypass/data_bypass_interface.h:
```diff
@@ -40,11 +42,13 @@ class RegisterStage
 
     static constexpr const Latency IN_RF_STAGE = Latency( MAX_VAL8);
     static constexpr const Latency MEM_STAGE   = IN_RF_STAGE - 2_lt;
+    static constexpr const Latency BRANCH_STAGE   = IN_RF_STAGE - 2_lt;
```
Please fix indentation here.

pavelkryukov commented on simulator/modules/decode/bypass/data_bypass.h:
```diff
@@ -159,6 +163,10 @@ void DataBypass<FuncInstr>::trace_new_dst_register( const Instr& instr, Register
     {
         entry.ready_stage.set_to_mem_stage();
     }
+    else if ( instr.is_branch())
```
`is_branch` or `is_jump`?

vodogrey2012 committed:
Solve logic problem

Exclude this->is_jump() from is_mem_stage_required() and add branch option to Latency get_instruction_latency( const Instr& instr)

vodogrey2012 committed:
Solve style problems and add if condition

pavelkryukov committed:
Merge branch 'master' into master

pavelkryukov committed:
Merge branch 'master' into master
